File: 496-next-greater-element-i/496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        int n=nums2.size();
        unordered_map<int,int>m;
        for(int i=0;i<n;i++){
            m[nums2[i]]=i;
        }
        stack<int>st;
        vector<int>result(n,-1);
        for(int i=0;i<n;i++){
            while(!st.empty() && nums2[st.top()]<nums2[i]){
                result[st.top()]=nums2[i];
                st.pop();
            }
            st.push(i);
        }
        vector<int>ans(nums1.size(),-1);
        for(int i=0;i<nums1.size();i++){
            ans[i]=result[m[nums1[i]]];
        }
        return ans;
    }
};

File: 496-next-greater-element-i/496-next-greater-element-i.ts
var nextGreaterElement = function(nums1, nums2) {
    let stack = [];
    let result = [];
    let map = new Map();
    
    stack.push(nums2[0]);
    
    for(let i = 1; i < nums2.length; i++){
        let current = nums2[i];
        
        while(stack.length > 0 && current > stack[stack.length - 1]){
            let top = stack.pop();
            map.set(top, current);
        }
        
        stack.push(current);
    }
    
    while(!(stack.length === 0)){
        let top = stack.pop();
        map.set(top, -1);
    }
   
    for(let i = 0; i < nums1.length; i++){
       let resVal = map.get(nums1[i]);
       result.push(resVal);
    }
    
    return result;
    
};